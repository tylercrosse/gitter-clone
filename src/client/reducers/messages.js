import moment from 'moment';
import { ADD_MESSAGE } from '../actions';

const burstify = (messages) => {
  const msgA = Object.values(messages);
  const bursts = {};
  let prev = msgA[0];
  prev.burstStart = true;
  bursts[prev.username + ':' + prev.createdAt] = [prev];
  for (let i = 1; i < msgA.length; i++) {
    const curr = msgA[i];
    const currTime = moment(curr.createdAt);
    const prevTime = moment(prev.createdAt);
    const rangeEnd = prevTime.clone().add(15, 'minutes');
    if (
      prev.username === curr.username &&
      currTime.isBetween(prevTime, rangeEnd, null, '[]')
    ) {
      curr.burstStart = false;
      bursts[prev.username + ':' + prev.createdAt].push(curr);
    } else {
      curr.burstStart = true;
      bursts[curr.username + ':' + curr.createdAt] = [curr];
      prev = curr;
    }
  }
  return Object.values(bursts).reduce((acc, cur) => acc.concat(cur));
};

const message = (state, action) => {
  switch (action.type) {
    case ADD_MESSAGE:
      return {
        [action.message._id]: {
          ...action.message
        }
      };
    default:
      return state;
  }
};

const messages = (state = {}, action) => {
  if (action.response && action.response.entities.messages) {
    const burstifiedMessages = burstify(action.response.entities.messages);
    return {
      ...state,
      ...burstifiedMessages
    };
  }
  switch (action.type) {
    case ADD_MESSAGE: {
      const burstifiedMessages = burstify({
        ...state,
        ...message(undefined, action)
      });
      return {
        ...state,
        ...burstifiedMessages
      };
    }
    default:
      return state;
  }
};

export default messages;
