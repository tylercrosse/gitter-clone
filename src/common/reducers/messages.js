import { normalize } from 'normalizr';
import { CALL_API,
  getJSON }          from 'redux-api-middleware';
import Schemas       from './schemas';
import { SIGN_OUT } from './user';

// actions
export const ADD_MESSAGE = 'ADD_MESSAGE';
export const MESSAGES_REQUEST = 'MESSAGES_REQUEST';
export const MESSAGES_SUCCESS = 'MESSAGES_SUCCESS';
export const MESSAGES_FAILURE = 'MESSAGES_FAILURE';

// action creators
export const addMessage = ({
  userId,
  text,
  rawMarkup,
  convo
}) => ({
  type: 'server.' + ADD_MESSAGE,
  userId,
  text,
  rawMarkup,
  convo
});

export const fetchMessages =  (convo) => ({
  [CALL_API]: {
    endpoint: window.location.origin + '/api/messages/' + convo,
    method: 'GET',
    types: [
      MESSAGES_REQUEST,
      {
        type: MESSAGES_SUCCESS,
        payload: /* istanbul ignore next */ (action, state, res) => {
          return getJSON(res)
          .then((json) => normalize(json, Schemas.MESSAGE_ARRAY));
        }
      },
      MESSAGES_FAILURE
    ]
  }
});


// reducers
export const message = (state, action) => {
  switch (action.type) {
    case ADD_MESSAGE:
      return {
        [action.message.createdAt]: {
          ...action.message
        }
      };
    default:
      return state;
  }
};

const messages = (state = {}, action) => {
  if (action.payload && action.payload.entities && action.payload.entities.messages) {
    return {
      ...state,
      ...action.payload.entities.messages
    };
  }
  switch (action.type) {
    case ADD_MESSAGE: {
      return {
        ...state,
        ...message(undefined, action)
      };
    }
    case SIGN_OUT:
      return {};
    default:
      return state;
  }
};

export default messages;
