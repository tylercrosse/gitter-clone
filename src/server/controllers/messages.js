import marked from 'marked';
import markedConfig from '../config/markedConfig';
import logger from '../config/logger';

const Message = require('../models').Message;
const Convo = require('../models').Convo;

marked.setOptions(markedConfig);

export const getMessages = (req, res) => {
  Convo.find({
    where: { name: req.params.convo },
    include: [
      {
        model: Message,
        as: 'messages'
      }
    ]
  })
    .then((convo) => {
      if (convo === null) res.status(404).send('Chat not found');
      else res.json(convo.messages);
    })
    .catch(
      /* istanbul ignore next */ (err) => {
        logger.log('error', err);
      }
    );
};

export const addMessage = (io, action) => {
  const rawMarkup = marked(action.text);
  const content = {
    username: action.username,
    text: action.text,
    rawMarkup
  };

  Convo.findOne({ where: { name: action.convo } })
    .then(/* istanbul ignore next */ (convo) => _createMessage(content, convo))
    .then(/* istanbul ignore next */ (result) => _emitAddMessage(io, result))
    .catch(
      /* istanbul ignore next */ (err) => {
        logger.log('error', err);
      }
    );
};

const _createMessage = (content, convo) =>
  Message.create({
    ...content,
    convoId: convo.id
  });
const _emitAddMessage = (io, message) =>
  io.emit('action', {
    // FIXME better decouple db & socket interactions
    type: 'ADD_MESSAGE',
    message
  });

export const pvt = { _createMessage, _emitAddMessage };
