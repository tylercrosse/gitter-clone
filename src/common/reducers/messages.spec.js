import messages, { message } from './messages';

describe('message reucer', () => {
  it('should handle ADD_MESSAGE', () => {
    expect(
      message({}, {
        type: 'ADD_MESSAGE',
        message: {
          createdAt: '2017-02-01T09:00:00-08:00',
          username: 'dan',
          text: 'Run the tests',
          id: 0
        }
      })
    ).toMatchSnapshot('-> 1 message');
  });

  it('should return state for default switch case', () => {
    expect(
      message({}, {
        type: 'fooBar'
      })
    ).toMatchSnapshot('default');
  });
});

describe('messages reducer', () => {
  it('should handle initial state', () => {
    expect(
      messages(undefined, {})
    ).toEqual({});
  });

  it('should return state for default switch case', () => {
    expect(
      messages({}, {
        type: 'fooBar'
      })
    ).toMatchSnapshot('default');
  });

  it('should handle fetch payload', () => {
    let action = {};
    expect(messages({}, action)).toMatchSnapshot('no payload');
    action = {
      payload: {
        entities: { messages: {}}
      }
    };
    expect(messages({}, action)).toMatchSnapshot('with payload');
  });

  it('should handle ADD_MESSAGE', () => {
    expect(
      messages({}, {
        type: 'ADD_MESSAGE',
        message: {
          createdAt: '2017-02-01T09:00:00-08:00',
          username: 'dan',
          text: 'Run the tests',
          id: 0
        }
      })
    ).toMatchSnapshot('-> 1 message');

    expect(
      messages({
        0: {
          createdAt: '2017-02-01T09:00:00-08:00',
          username: 'dan',
          text: 'Run the tests',
          id: 0
        }
      }, {
        type: 'ADD_MESSAGE',
        message: {
          // 1 min later, same burst
          createdAt: '2017-02-01T09:01:00-08:00',
          username: 'dan',
          text: 'Use Redux',
          id: 1
        }
      })
    ).toMatchSnapshot('-> 2 messages');

    expect(
      messages({
        0: {
          createdAt: '2017-02-01T09:00:00-08:00',
          username: 'dan',
          text: 'Run the tests',
          id: 0
        },
        1: {
          // 1 min later, same burst
          createdAt: '2017-02-01T09:01:00-08:00',
          username: 'dan',
          text: 'Use Redux',
          id: 1
        }
      }, {
        type: 'ADD_MESSAGE',
        message: {
           // 20 min later, new burst
          createdAt: '2017-02-01T09:20:00-08:00',
          username: 'dan',
          text: 'Fix the tests',
          id: 2
        }
      })
    ).toMatchSnapshot('-> 3 messages');
  });
});
