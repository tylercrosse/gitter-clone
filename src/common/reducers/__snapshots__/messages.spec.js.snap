// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`-> 1 message 1`] = `
Object {
  "2017-02-01T09:00:00-08:00": Object {
    "createdAt": "2017-02-01T09:00:00-08:00",
    "id": 0,
    "text": "Run the tests",
    "username": "dan",
  },
}
`;

exports[`-> 1 message 2`] = `
Object {
  "2017-02-01T09:00:00-08:00": Object {
    "createdAt": "2017-02-01T09:00:00-08:00",
    "id": 0,
    "text": "Run the tests",
    "username": "dan",
  },
}
`;

exports[`-> 2 messages 1`] = `
Object {
  "0": Object {
    "createdAt": "2017-02-01T09:00:00-08:00",
    "id": 0,
    "text": "Run the tests",
    "username": "dan",
  },
  "2017-02-01T09:01:00-08:00": Object {
    "createdAt": "2017-02-01T09:01:00-08:00",
    "id": 1,
    "text": "Use Redux",
    "username": "dan",
  },
}
`;

exports[`-> 3 messages 1`] = `
Object {
  "0": Object {
    "createdAt": "2017-02-01T09:00:00-08:00",
    "id": 0,
    "text": "Run the tests",
    "username": "dan",
  },
  "1": Object {
    "createdAt": "2017-02-01T09:01:00-08:00",
    "id": 1,
    "text": "Use Redux",
    "username": "dan",
  },
  "2017-02-01T09:20:00-08:00": Object {
    "createdAt": "2017-02-01T09:20:00-08:00",
    "id": 2,
    "text": "Fix the tests",
    "username": "dan",
  },
}
`;

exports[`default 1`] = `Object {}`;

exports[`default 2`] = `Object {}`;

exports[`no payload 1`] = `Object {}`;

exports[`with payload 1`] = `Object {}`;
